enterprise/metrics/src/main/java/org/neo4j/metrics/output/CsvReporterSingle.java

```diff
@@ -0,0 +1,413 @@
+    private final Clock clock;
+    private final Log logger;
+    private PrintWriter out;
+
+    private CsvReporterSingle( MetricRegistry registry,
+                               File file,
+                               TimeUnit rateUnit,
+                               TimeUnit durationUnit,
+                               Clock clock,
+                               Log logger,
+                               MetricFilter filter )
+    {
+        super( registry, "csv-reporter-single", filter, rateUnit, durationUnit );
+        this.file = file;
+        this.clock = clock;
+        this.logger = logger;
+    }
+
+    @Override
+    public void stop()
+    {
+        super.stop();
+
+        if ( out != null )
+        {
+            out.close();
+        }
+    }
+
+    @Override
+    public void report( SortedMap<String,Gauge> gauges,
+            SortedMap<String,Counter> counters,
+            SortedMap<String,Histogram> histograms,
+            SortedMap<String,Meter> meters,
+            SortedMap<String,Timer> timers )
+    {
+        if ( out == null )
+        {
+            try
+            {
+                startNewCsvFile( gauges, counters, histograms, meters, timers );
+            }
+            catch ( Exception e )
+            {
+                logger.warn( "Could not create output CSV file: " + file.getAbsolutePath(), e );
+            }
+        }
+
+        StringBuilder line = new StringBuilder();
+
+        final long timestamp = TimeUnit.MILLISECONDS.toSeconds( clock.getTime() );
+
+        line.append( timestamp ).append( SEPARATOR ).append( ISO8601.get().format( new Date( clock.getTime() ) ) );
+
+        for ( Map.Entry<String,Gauge> entry : gauges.entrySet() )
+        {
+            reportGauge( entry.getValue(), line );
+        }
+
+        for ( Map.Entry<String,Counter> entry : counters.entrySet() )
+        {
+            reportCounter( entry.getValue(), line );
+        }
+
+        for ( Map.Entry<String,Histogram> entry : histograms.entrySet() )
+        {
+            reportHistogram( entry.getValue(), line );
+        }
+
+        for ( Map.Entry<String,Meter> entry : meters.entrySet() )
+        {
+            reportMeter( entry.getValue(), line );
+        }
+
+        for ( Map.Entry<String,Timer> entry : timers.entrySet() )
+        {
+            reportTimer( entry.getValue(), line );
+        }
+
+        out.println( line.toString() );
+        out.flush();
+    }
+
+    private void startNewCsvFile( SortedMap<String,Gauge> gauges, SortedMap<String,Counter> counters,
+                                  SortedMap<String,Histogram> histograms, SortedMap<String,Meter> meters,
+                                  SortedMap<String,Timer> timers ) throws FileNotFoundException
+    {
+        if ( file.exists() )
+        {
+            archiveExistingLogFile();
+        }
+
+        out = new PrintWriter( new OutputStreamWriter( new FileOutputStream( file ), StandardCharsets.UTF_8 ) );
```

Comment:
Super annoying that there's no constructor that takes the Charset instance directly.